File: src/main/java/io/github/fisher2911/hmcleaves/packet/BlockBreakManager.java
/*
 *
 *  *     HMCLeaves
 *  *     Copyright (C) 2022  Hibiscus Creative Studios
 *  *
 *  *     This program is free software: you can redistribute it and/or modify
 *  *     it under the terms of the GNU General Public License as published by
 *  *     the Free Software Foundation, either version 3 of the License, or
 *  *     (at your option) any later version.
 *  *
 *  *     This program is distributed in the hope that it will be useful,
 *  *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  *     GNU General Public License for more details.
 *  *
 *  *     You should have received a copy of the GNU General Public License
 *  *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

package io.github.fisher2911.hmcleaves.packet;

import io.github.fisher2911.hmcleaves.HMCLeaves;
import io.github.fisher2911.hmcleaves.cache.BlockCache;
import io.github.fisher2911.hmcleaves.config.LeavesConfig;
import io.github.fisher2911.hmcleaves.data.BlockData;
import io.github.fisher2911.hmcleaves.data.LogData;
import io.github.fisher2911.hmcleaves.data.MineableData;
import io.github.fisher2911.hmcleaves.util.ItemUtil;
import io.github.fisher2911.hmcleaves.world.Position;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.bukkit.scheduler.BukkitTask;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.SplittableRandom;
import java.util.UUID;
import java.util.function.Supplier;

public class BlockBreakManager {

    private static final SplittableRandom RANDOM = new SplittableRandom();

    private final HMCLeaves plugin;
    private final BlockCache blockCache;
    private final LeavesConfig leavesConfig;
    private final Map<UUID, BlockBreakData> blockBreakDataMap;


    public BlockBreakManager(Map<UUID, BlockBreakData> blockBreakDataMap, HMCLeaves plugin) {
        this.blockBreakDataMap = blockBreakDataMap;
        this.plugin = plugin;
        this.blockCache = this.plugin.getBlockCache();
        this.leavesConfig = this.plugin.getLeavesConfig();
    }

    @Nullable
    public BlockBreakData getBlockBreakData(UUID uuid) {
        return this.blockBreakDataMap.get(uuid);
    }

    private static final int LOG_HARDNESS = 2;
    public static final BlockBreakModifier LOG_BREAK_MODIFIER = new BlockBreakModifier(LOG_HARDNESS, false, Set.of(BlockBreakModifier.ToolType.AXE));

    public <T extends MineableData & BlockData> void startBlockBreak(Player player, Position position, T mineableData) {
        if (mineableData.blockBreakModifier() == null) return;
        final int blockBreakTime = this.calculateBlockBreakTimeInTicks(
                player,
                player.getInventory().getItemInMainHand(),
                mineableData.blockBreakModifier()
        );
//        final int period = (int) Math.ceil(blockBreakTime / (double) BlockBreakData.MAX_DAMAGE);
        final BlockBreakData blockBreakData = new BlockBreakData(
                RANDOM.nextInt(10_000, 20_000),
                mineableData,
                player,
                position,
                blockBreakTime,
                0,
                this.createScheduler(blockBreakTime, player.getUniqueId())
        );
        this.blockBreakDataMap.put(player.getUniqueId(), blockBreakData);
    }

    public void cancelBlockBreak(Player player) {
        final BlockBreakData blockBreakData = this.blockBreakDataMap.get(player.getUniqueId());
        if (blockBreakData == null) {
            return;
        }
        PacketUtils.sendBlockBreakAnimation(player, blockBreakData.getPosition().toLocation(), blockBreakData.getEntityId(), (byte) -1);
        blockBreakData.getBreakTask().cancel();
        this.blockBreakDataMap.remove(player.getUniqueId());
    }

    private BukkitTask createScheduler(int blockBreakTime, UUID uuid) {
        return Bukkit.getScheduler().runTaskTimer(this.plugin,
                () -> {
                    final BlockBreakData blockBreakData = BlockBreakManager.this.blockBreakDataMap.get(uuid);
                    if (blockBreakData == null) {
                        return;
                    }
                    final Player player = blockBreakData.getBreaker();
                    if (blockBreakData.isBroken()) {
                        final Block block = blockBreakData.getPosition().toLocation().getBlock();
                        final LeavesBlockBreakEvent event = new LeavesBlockBreakEvent(block, player);
                        Bukkit.getPluginManager().callEvent(event);
                        if (event.isCancelled()) {
                            blockBreakData.resetProgress();
                            return;
                        }
                        PacketUtils.sendBlockBreakAnimation(
                                blockBreakData.getBreaker(),
                                blockBreakData.getPosition().toLocation(),
                                blockBreakData.getEntityId(),
                                (byte) -1
                        );
                        PacketUtils.sendBlockBroken(
                                player,
                                blockBreakData.getPosition(),
                                blockBreakData.getBlockData().sendBlockId()
                        );
                        BlockBreakManager.this.blockCache.removeBlockData(blockBreakData.getPosition());
                        block.setType(Material.AIR);
                        Supplier<ItemStack> itemStackSupplier = null;
                        final BlockBreakModifier blockBreakModifier = blockBreakData.getBlockData().blockBreakModifier();
                        final ItemStack heldItem = blockBreakData.getBreaker().getInventory().getItemInMainHand();
                        if (!blockBreakModifier.requiresToolToDrop() || blockBreakModifier.hasToolType(heldItem.getType())) {
                            if (blockBreakData.getBlockData() instanceof final LogData logData) {
                                itemStackSupplier = BlockBreakManager.this.leavesConfig.getItemSupplier(logData.getCurrentId());
                            } else {
                                itemStackSupplier = BlockBreakManager.this.leavesConfig.getItemSupplier(blockBreakData.getBlockData().id());
                            }
                            if (itemStackSupplier == null) {
                                itemStackSupplier = () -> new ItemStack(blockBreakData.getBlockData().worldBlockType());
                            }
                        }
                        if (event.isDropItems() && itemStackSupplier != null) {
                            final World world = block.getWorld();
                            final ItemStack itemStack = itemStackSupplier.get();
                            if (itemStack != null) {
                                final Supplier<ItemStack> finalItemStackSupplier = itemStackSupplier;
                                Bukkit.getScheduler().runTaskLater(BlockBreakManager.this.plugin, () -> world.dropItem(block.getLocation().clone().add(0.5, 0, 0.5), finalItemStackSupplier.get()), 1);
                            }
                        }
                        blockBreakData.getBreakTask().cancel();
                        if (!ItemUtil.isQuickMiningTool(heldItem.getType())) {
                            PacketUtils.removeMiningFatigue(player);
                        }
                        BlockBreakManager.this.blockBreakDataMap.remove(uuid);
                        return;
                    }
                    final int updatedBlockBreakTime = BlockBreakManager.this.calculateBlockBreakTimeInTicks(
                            player,
                            player.getInventory().getItemInMainHand(),
                            blockBreakData.getBlockData().blockBreakModifier()
                    );
                    final double percentageToAdd = (double) blockBreakTime / updatedBlockBreakTime;
                    blockBreakData.addBreakTimeProgress(percentageToAdd);
//                        blockBreakData.addBreakTimeProgress(period);
                    blockBreakData.send(blockBreakData.getPosition());
                }, 1, 1);
    }

    private static final Map<Material, Integer> TOOL_SPEED_MULTIPLIERS = createToolSpeedModifiers();

    // formula found at https://minecraft.fandom.com/wiki/Breaking#Speed
    private int calculateBlockBreakTimeInTicks(Player player, ItemStack inHand, BlockBreakModifier modifier) {
        final Material material = inHand.getType();
        double speedMultiplier = TOOL_SPEED_MULTIPLIERS.getOrDefault(material, 1);
        if (!modifier.hasToolType(material)) {
            speedMultiplier = 1;
        }
        final int efficiencyLevel = inHand.getEnchantmentLevel(Enchantment.DIG_SPEED);
        if (efficiencyLevel != 0) {
            speedMultiplier += Math.pow(efficiencyLevel, 2) + 1;
        }
        final PotionEffect haste = player.getPotionEffect(PotionEffectType.FAST_DIGGING);
        if (haste != null) {
            speedMultiplier *= 0.2 * haste.getAmplifier() + 1;
        }
        final PotionEffect miningFatigue = player.getPotionEffect(PotionEffectType.SLOW_DIGGING);
        if (miningFatigue != null) {
            speedMultiplier *= Math.pow(0.3, Math.min(miningFatigue.getAmplifier(), 4));
        }
        final ItemStack helmet = player.getInventory().getHelmet();
        if (player.isInWater()) {
            if (helmet == null || helmet.getEnchantmentLevel(Enchantment.WATER_WORKER) <= 0) {
                speedMultiplier /= 5;
            }
        }
        if (!player.isOnGround()) {
            speedMultiplier /= 5;
        }
        final double damage = (speedMultiplier / modifier.hardness()) / 30;
        if (damage > 1) {
            return 0;
        }
        return (int) Math.ceil(1 / damage);
    }

    private static Map<Material, Integer> createToolSpeedModifiers() {
        final Map<Material, Integer> map = new HashMap<>();
        final Set<Material> goldTools = Set.of(Material.GOLDEN_SHOVEL, Material.GOLDEN_PICKAXE, Material.GOLDEN_AXE, Material.GOLDEN_HOE, Material.GOLDEN_SWORD);
        final Set<Material> netheriteTools = Set.of(Material.NETHERITE_SHOVEL, Material.NETHERITE_PICKAXE, Material.NETHERITE_AXE, Material.NETHERITE_HOE, Material.NETHERITE_SWORD);
        final Set<Material> diamondTools = Set.of(Material.DIAMOND_SHOVEL, Material.DIAMOND_PICKAXE, Material.DIAMOND_AXE, Material.DIAMOND_HOE, Material.DIAMOND_SWORD);
        final Set<Material> ironTools = Set.of(Material.IRON_SHOVEL, Material.IRON_PICKAXE, Material.IRON_AXE, Material.IRON_HOE, Material.IRON_SWORD);
        final Set<Material> stoneTools = Set.of(Material.STONE_SHOVEL, Material.STONE_PICKAXE, Material.STONE_AXE, Material.STONE_HOE, Material.STONE_SWORD);
        final Set<Material> woodenTools = Set.of(Material.WOODEN_SHOVEL, Material.WOODEN_PICKAXE, Material.WOODEN_AXE, Material.WOODEN_HOE, Material.WOODEN_SWORD);
        goldTools.forEach(material -> map.put(material, 12));
        netheriteTools.forEach(material -> map.put(material, 9));
        diamondTools.forEach(material -> map.put(material, 8));
        ironTools.forEach(material -> map.put(material, 6));
        stoneTools.forEach(material -> map.put(material, 4));
        woodenTools.forEach(material -> map.put(material, 2));
        map.put(Material.SHEARS, 2);
        return map;
    }

//    public void setBlockBreakData(UUID uuid, BlockBreakData blockBreakData) {
//        this.blockBreakDataMap.put(uuid, blockBreakData);
//    }

    private static class BlockBreakData {

        public static final byte MAX_DAMAGE = 10;

        private final int entityId;
        private final BlockData blockData;
        private final Player breaker;
        private final Position position;
        private final int totalBreakTime;
        private double breakTimeProgress;
        private final BukkitTask breakTask;

        public <T extends BlockData & MineableData> BlockBreakData(
                int entityId,
                T blockData,
                Player breaker,
                Position position,
                int totalBreakTime,
                int breakTimeProgress,
                BukkitTask breakTask
        ) {
            this.entityId = entityId;
            this.blockData = blockData;
            this.breaker = breaker;
            this.position = position;
            this.totalBreakTime = totalBreakTime;
            this.breakTimeProgress = breakTimeProgress;
            this.breakTask = breakTask;
        }

        public int getEntityId() {
            return entityId;
        }

        public <T extends BlockData & MineableData> T getBlockData() {
            return (T) this.blockData;
        }

        public Player getBreaker() {
            return breaker;
        }

        public Position getPosition() {
            return position;
        }

        public int getTotalBreakTime() {
            return totalBreakTime;
        }

        public double getBreakTimeProgress() {
            return breakTimeProgress;
        }

        public void setBreakTimeProgress(double breakTimeProgress) {
            this.breakTimeProgress = breakTimeProgress;
        }

        public void addBreakTimeProgress(double breakTimeProgress) {
            this.breakTimeProgress = Math.min(this.breakTimeProgress + breakTimeProgress, this.totalBreakTime);
        }

        public byte calculateDamage() {
            final double percentage = (double) this.breakTimeProgress / this.totalBreakTime;
            final double damage = MAX_DAMAGE * percentage;
            return (byte) (Math.min(damage, MAX_DAMAGE) - 1);
        }

        public void send(Position position) {
            final byte calculatedDamage = this.calculateDamage();
            PacketUtils.sendBlockBreakAnimation(this.breaker, position.toLocation(), this.entityId, calculatedDamage);
        }

        public boolean isBroken() {
            return this.breakTimeProgress >= this.totalBreakTime;
        }

        public BukkitTask getBreakTask() {
            return breakTask;
        }

        public void resetProgress() {
            this.breakTimeProgress = 0;
        }

    }

    public static class LeavesBlockBreakEvent extends BlockBreakEvent {

        public LeavesBlockBreakEvent(@NotNull Block theBlock, @NotNull Player player) {
            super(theBlock, player);
        }

    }

}

File: common/src/main/kotlin/com/hibiscusmc/hmcleaves/packet/mining/BlockBreakManager.kt
package com.hibiscusmc.hmcleaves.packet.mining

import com.hibiscusmc.hmcleaves.HMCLeaves
import com.hibiscusmc.hmcleaves.block.BlockData
import com.hibiscusmc.hmcleaves.config.LeavesConfig
import com.hibiscusmc.hmcleaves.packet.removeMiningFatigue
import com.hibiscusmc.hmcleaves.packet.sendBlockBreakAnimation
import com.hibiscusmc.hmcleaves.packet.sendBlockBroken
import com.hibiscusmc.hmcleaves.world.Position
import com.hibiscusmc.hmcleaves.world.WorldManager
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffectType
import org.bukkit.scheduler.BukkitTask
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer
import kotlin.math.ceil
import kotlin.math.min
import kotlin.math.pow


class BlockBreakManager(
    private val blockBreakDataMap: MutableMap<UUID, BlockBreakData> = ConcurrentHashMap(),
    private val plugin: HMCLeaves,
    private val worldManager: WorldManager = plugin.worldManager,
    private val config: LeavesConfig = plugin.leavesConfig
) {
    private val randomRange = (10_000..20_000)

    fun getBlockBreakData(uuid: UUID): BlockBreakData? {
        return blockBreakDataMap[uuid]
    }

    fun startBlockBreak(
        player: Player,
        position: Position,
        blockData: BlockData
    ) {
        if (blockData.blockBreakModifier == null) return
        val blockBreakTime = this.calculateBlockBreakTimeInTicks(
            player,
            player.inventory.itemInMainHand,
            blockData.blockBreakModifier
        )
        val blockBreakData = BlockBreakData(
            randomRange.random(),
            blockData,
            player,
            position,
            blockBreakTime,
            0.0,
            this.createScheduler(blockBreakTime, player.uniqueId)
        )
        blockBreakDataMap[player.uniqueId] = blockBreakData
    }

    fun cancelBlockBreak(player: Player) {
        val blockBreakData = blockBreakDataMap[player.uniqueId] ?: return
        sendBlockBreakAnimation(
            player,
            blockBreakData.position,
            blockBreakData.entityId,
            (-1).toByte()
        )
        blockBreakData.breakTask.cancel()
        blockBreakDataMap.remove(player.uniqueId)
    }

    private fun createScheduler(blockBreakTime: Int, uuid: UUID): BukkitTask {
        return Bukkit.getScheduler().runTaskTimer(
            this.plugin,
            Runnable {
                val blockBreakData = blockBreakDataMap[uuid] ?: return@Runnable
                val player = blockBreakData.breaker
                if (blockBreakData.isBroken) {
                    val position = blockBreakData.position
                    val block: Block = position.toLocation().block
                    val event = BlockBreakEvent(block, player)
                    Bukkit.getPluginManager().callEvent(event)
                    if (event.isCancelled) {
                        blockBreakData.resetProgress()
                        return@Runnable
                    }
                    Bukkit.getScheduler().runTaskAsynchronously(this.plugin, Runnable {
                        sendBlockBreakAnimation(
                            blockBreakData.breaker,
                            position,
                            blockBreakData.entityId,
                            (-1).toByte()
                        )
                        sendBlockBroken(
                            player,
                            position,
                            blockBreakData.blockData.getBlockGlobalId()
                        )
                    })
                    worldManager[position.world]
                        ?.get(position.getChunkPosition())
                        ?.remove(position.toPositionInChunk(), false)
                    block.type = Material.AIR
                    val blockBreakModifier = blockBreakData.blockData.blockBreakModifier!!
                    val heldItem: ItemStack = blockBreakData.breaker.inventory.itemInMainHand
                    val dropItems = blockBreakModifier.hasEnchantment(heldItem) &&
                            (!blockBreakModifier.requiresToolToDrop ||
                                    blockBreakModifier.hasToolType(heldItem.type))

                    if (event.isDropItems && dropItems) {
                        val world: World = block.world
                        Bukkit.getScheduler().runTaskLater(
                            this@BlockBreakManager.plugin,
                            Runnable {
                                for (itemStack in blockBreakData.blockData.getDrops(config)) {
                                    world.dropItem(
                                        block.location.clone().add(0.5, 0.0, 0.5),
                                        itemStack
                                    )
                                }
                            }, 1
                        )

                    }
                    removeMiningFatigue(player);
                    blockBreakData.breakTask.cancel()
                    blockBreakDataMap.remove(uuid)
                    return@Runnable
                }
                val updatedBlockBreakTime =
                    calculateBlockBreakTimeInTicks(
                        player,
                        player.inventory.itemInMainHand,
                        blockBreakData.blockData.blockBreakModifier!!
                    )
                val percentageToAdd = blockBreakTime.toDouble() / updatedBlockBreakTime
                blockBreakData.addBreakTimeProgress(percentageToAdd)
                blockBreakData.send(blockBreakData.position)
            }, 1, 1
        )
    }

    // formula found at https://minecraft.fandom.com/wiki/Breaking#Speed
    private fun calculateBlockBreakTimeInTicks(player: Player, inHand: ItemStack, modifier: BlockBreakModifier): Int {
        val material: Material = inHand.type
        var speedMultiplier = TOOL_SPEED_MULTIPLIERS.getOrDefault(material, 1).toDouble()
        if (!modifier.hasToolType(material)) {
            speedMultiplier = 1.0
        }
        val efficiencyLevel: Int = inHand.getEnchantmentLevel(Enchantment.DIG_SPEED)
        if (efficiencyLevel != 0) {
            speedMultiplier += efficiencyLevel.toDouble().pow(2.0) + 1
        }
        val haste = player.getPotionEffect(PotionEffectType.FAST_DIGGING)
        if (haste != null) {
            speedMultiplier *= 0.2 * haste.amplifier + 1
        }
        val miningFatigue = player.getPotionEffect(PotionEffectType.SLOW_DIGGING)
        if (miningFatigue != null) {
            speedMultiplier *= 0.3.pow(min(miningFatigue.amplifier.toDouble(), 4.0))
        }
        val helmet: ItemStack? = player.inventory.helmet
        if (player.isInWater) {
            if (helmet == null || helmet.getEnchantmentLevel(Enchantment.WATER_WORKER) <= 0) {
                speedMultiplier /= 5.0
            }
        }
        if (!player.isOnGround) {
            speedMultiplier /= 5.0
        }
        val damage: Double = (speedMultiplier / modifier.hardness) / 30
        if (damage > 1) {
            return 0
        }
        return ceil(1 / damage).toInt()
    }

    class BlockBreakData(
        val entityId: Int,
        val blockData: BlockData,
        val breaker: Player,
        val position: Position,
        private val totalBreakTime: Int,
        private var breakTimeProgress: Double,
        val breakTask: BukkitTask
    ) {
        fun addBreakTimeProgress(breakTimeProgress: Double) {
            this.breakTimeProgress = min(
                this.breakTimeProgress + breakTimeProgress,
                totalBreakTime.toDouble()
            )
        }

        private fun calculateDamage(): Byte {
            val percentage = breakTimeProgress / this.totalBreakTime
            val damage = MAX_DAMAGE * percentage
            return (min(damage, MAX_DAMAGE.toDouble()) - 1).toInt().toByte()
        }

        fun send(position: Position) {
            val calculatedDamage = this.calculateDamage()
            sendBlockBreakAnimation(this.breaker, position, this.entityId, calculatedDamage)
        }

        val isBroken: Boolean
            get() = this.breakTimeProgress >= this.totalBreakTime

        fun resetProgress() {
            this.breakTimeProgress = 0.0
        }

        companion object {
            const val MAX_DAMAGE: Byte = 10
        }
    }

    companion object {

        private const val LOG_HARDNESS = 2.0
        val LOG_BREAK_MODIFIER: BlockBreakModifier = BlockBreakModifier(
            LOG_HARDNESS,
            false,
            mutableSetOf(ToolType.AXE),
            setOf()
        )

        private val TOOL_SPEED_MULTIPLIERS = createToolSpeedModifiers()

        private fun createToolSpeedModifiers(): Map<Material, Int> {
            val map: MutableMap<Material, Int> = EnumMap(org.bukkit.Material::class.java)
            val goldTools = EnumSet.of(
                Material.GOLDEN_SHOVEL,
                Material.GOLDEN_PICKAXE,
                Material.GOLDEN_AXE,
                Material.GOLDEN_HOE,
                Material.GOLDEN_SWORD
            )
            val netheriteTools = EnumSet.of(
                Material.NETHERITE_SHOVEL,
                Material.NETHERITE_PICKAXE,
                Material.NETHERITE_AXE,
                Material.NETHERITE_HOE,
                Material.NETHERITE_SWORD
            )
            val diamondTools = EnumSet.of(
                Material.DIAMOND_SHOVEL,
                Material.DIAMOND_PICKAXE,
                Material.DIAMOND_AXE,
                Material.DIAMOND_HOE,
                Material.DIAMOND_SWORD
            )
            val ironTools = EnumSet.of(
                Material.IRON_SHOVEL,
                Material.IRON_PICKAXE,
                Material.IRON_AXE,
                Material.IRON_HOE,
                Material.IRON_SWORD
            )
            val stoneTools = EnumSet.of(
                Material.STONE_SHOVEL,
                Material.STONE_PICKAXE,
                Material.STONE_AXE,
                Material.STONE_HOE,
                Material.STONE_SWORD
            )
            val woodenTools = EnumSet.of(
                Material.WOODEN_SHOVEL,
                Material.WOODEN_PICKAXE,
                Material.WOODEN_AXE,
                Material.WOODEN_HOE,
                Material.WOODEN_SWORD
            )
            goldTools.forEach(Consumer { material: Material -> map[material] = 12 })
            netheriteTools.forEach(Consumer { material: Material -> map[material] = 9 })
            diamondTools.forEach(Consumer { material: Material -> map[material] = 8 })
            ironTools.forEach(Consumer { material: Material -> map[material] = 6 })
            stoneTools.forEach(Consumer { material: Material -> map[material] = 4 })
            woodenTools.forEach(Consumer { material: Material -> map[material] = 2 })
            map[Material.SHEARS] = 2
            return map
        }
    }
}